From a267031f384a4433fdcd662a97bce7c4949d3fd6 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date: Fri, 5 Apr 2019 17:23:15 +0100
Subject: [PATCH] drm: vc4: Select display to blank during
 initialisation

Otherwise the rainbow splash screen remained in the display list

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>
---
 drivers/gpu/drm/vc4/vc4_firmware_kms.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

--- a/drivers/gpu/drm/vc4/vc4_firmware_kms.c
+++ b/drivers/gpu/drm/vc4/vc4_firmware_kms.c
@@ -87,6 +87,13 @@ struct fb_alloc_tags {
 	u32 layer;
 };
 
+struct mailbox_blank_display {
+	struct rpi_firmware_property_tag_header tag1;
+	u32 display;
+	struct rpi_firmware_property_tag_header tag2;
+	u32 blank;
+};
+
 static const struct vc_image_format {
 	u32 drm;	/* DRM_FORMAT_* */
 	u32 vc_image;	/* VC_IMAGE_* */
@@ -854,7 +861,12 @@ static int vc4_fkms_create_screen(struct
 	struct drm_crtc *crtc;
 	struct drm_plane *primary_plane, *overlay_plane, *cursor_plane;
 	struct drm_plane *destroy_plane, *temp;
-	u32 blank = 1;
+	struct mailbox_blank_display blank = {
+		.tag1 = {RPI_FIRMWARE_FRAMEBUFFER_SET_DISPLAY_NUM, 4, 0, },
+		.display = display_idx,
+		.tag2 = { RPI_FIRMWARE_FRAMEBUFFER_BLANK, 4, 0, },
+		.blank = 1,
+	};
 	int ret;
 
 	vc4_crtc = devm_kzalloc(dev, sizeof(*vc4_crtc), GFP_KERNEL);
@@ -865,9 +877,7 @@ static int vc4_fkms_create_screen(struct
 	vc4_crtc->display_number = display_ref;
 
 	/* Blank the firmware provided framebuffer */
-	rpi_firmware_property(vc4->firmware,
-			      RPI_FIRMWARE_FRAMEBUFFER_BLANK,
-			      &blank, sizeof(blank));
+	rpi_firmware_property_list(vc4->firmware, &blank, sizeof(blank));
 
 	primary_plane = vc4_fkms_plane_init(drm, DRM_PLANE_TYPE_PRIMARY,
 					    display_ref,
